core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -211,7 +211,13 @@ public TransportClientNodesService removeTransportAddress(TransportAddress trans
     }
 
     public <Response> void execute(NodeListenerCallback<Response> callback, ActionListener<Response> listener) {
-        List<DiscoveryNode> nodes = this.nodes;
+        List<DiscoveryNode> nodes;
+        synchronized (mutex) {
```

Comment:
> we know that the close() method first sets `close` to true and then empties the nodes list.

Sorry, that's relying too much on the implementation rather than on the invariants of the method.